composer/datasets/streaming/writer.py

```diff
@@ -0,0 +1,101 @@
+import os
+from typing import Dict, Iterable, List, Optional
+
+from tqdm import tqdm
+
+from composer.datasets.streaming.format import (StreamingDatasetIndex, get_index_basename, get_shard_basename,
+                                                sample_dict_to_bytes)
+
+
+class StreamingDatasetWriter(object):
+    """Writes StreamingDatasets."""
+
+    def __init__(self, dirname: str, fields: List[str], shard_size_limit: int = 1 << 24) -> None:
+        """Initialize with the given output dirname.
+
+        Args:
+            dirname (str): Directory to write shards to.
+            fields: (list of str): The fields to save for each sample.
+            shard_size_limit (int, optional): Maximum shard size in bytes. Default: 1 << 24.
+        """
+        assert len(fields) == len(set(fields))
+        assert 1 <= shard_size_limit
+
+        self.dirname = dirname
+        self.fields = fields
+        self.shard_size_limit = shard_size_limit
+
+        # Stats about shards written so far.
+        self.samples_per_shard = []
+        self.bytes_per_shard = []
+        self.bytes_per_sample = []
+
+        # Data of the shard in progress.
+        self.new_samples = []
+        self.new_shard_size = 0
+
+    def _flush_shard(self) -> None:
+        """Flush cached samples to a new dataset shard."""
+        if not self.samples_per_shard:
+            os.makedirs(self.dirname)
+        shard = len(self.samples_per_shard)
+        basename = get_shard_basename(shard)
+        filename = os.path.join(self.dirname, basename)
```

Comment:
Is it possible to write first to a temp file, and them rename upon successful completion, so that the writing can be gracefully resumed in case if it crashes?

Reply:
I'd rather have it die very obviously than risk unnoticed subtle differences in a continued run to write shards?